### PR TITLE
Fix Issue 368 Spider API URL List

### PR DESCRIPTION
**Pull request by kingthorin**
Added a Spider API end point:
- allUrls: Returns a list of all URLs found by all spider scans (based
on HistoryReferences)

Fixes zaproxy/zaproxy#368

### CODE REVIEW

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 1]
```diff
@@ -138,6 +140,8 @@ public SpiderAPI(ExtensionSpider extension) {
 		this.addApiView(new ApiView(VIEW_FULL_RESULTS, new String[] { PARAM_SCAN_ID }));
 		this.addApiView(new ApiView(VIEW_SCANS));
 		this.addApiView(new ApiView(VIEW_EXCLUDED_FROM_SCAN));
+		this.addApiView(new ApiView(VIEW_URLS, new String[] { PARAM_SCAN_ID }));
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: This view seems to duplicate `VIEW_RESULTS` (but depends on following comments).

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 2]
```diff
@@ -460,6 +464,26 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
 				resultList.addItem(new ApiResponseSet("scan", map));
 			}
 			result = resultList;
+		} else if (VIEW_URLS.equals(name)) {
+			ApiResponseList resultUrls = new ApiResponseList(name);
+			SpiderScan scan = (SpiderScan) this.getSpiderScan(params);
+			synchronized (scan.getResourcesFound()) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: This can be extracted to a method and used by both views.

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 3]
```diff
@@ -460,6 +464,26 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+				for (SpiderResource sr : scan.getResourcesFound()) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Should also include URLs out of scope? (As those were also found by the spider.)
Should probably add the URLs to a Set (to prevent duplicate URLs).

**Comment by thc202** [comment]
The view `allUrls` would be more useful if the scan results were kept indefinitely (they are not persisted between sessions and can be removed/discarded by the user).
Maybe it should use the persisted (spider) messages to get all URLs fetched by all spiders? That would give more consistent results.

**Comment by kingthorin** [comment]
So I should base it on history references where type is spider or spider_task?

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java (hunk 3 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: I guess it could build two sets kind of like fullResults (VIEW_FULL_RESULTS) does...

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java (hunk 2 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Sure if we keep 2.

**Comment by thc202** [comment]
Including both types should be OK (we might need to take care of duplicates though).

**Comment by kingthorin** [comment]
@psiinon #368 was originally opened by you. Any feedback? Based on what's here I guess I'll drop foundUrls (since it duplicates something that's there already). Then make allUrls based on history (using a set to avoid duplicates). Will that fill the original need?

**Comment by kingthorin** [comment]
Made the executive decision to proceed as outlined :smiley: Let me know what you think.

**Comment by psiinon** [comment]
I'm good with that :)

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 4]
```diff
@@ -460,6 +466,29 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+
+			if (extHistory == null) {
+				throw new ApiException(Type.NO_IMPLEMENTOR, ExtensionHistory.NAME);
+			}
+	        
+			ApiResponseList resultUrls = new ApiResponseList(name);
+			Set<String> urlSet=new HashSet<String>();
+			
+			for (int idx = 0; idx <= extHistory.getLastHistoryId(); idx++) {
+				HistoryReference hr=extHistory.getHistoryReference(idx);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: This requires to read (and caching almost) all HTTP messages, even if they are not of interest. It could obtain the spider HTTP messages directly, for example:
```Java
TableHistory tableHistory = extension.getModel().getDb().getTableHistory();
List<Integer> ids = tableHistory.getHistoryIdsOfHistType(extension.getModel().getSession().getSessionId(), HistoryReference.TYPE_SPIDER, HistoryReference.TYPE_SPIDER_TASK);
for (Integer id : ids) {
    String uri = tableHistory.read(id.intValue()).getHttpMessage().getRequestHeader().getURI().toString();
    // ...
```

**Comment by thc202** [comment]
The view could have a description (i.e. using resource message key "spider.api.view.allUrls").

**Comment by kingthorin** [comment]
I was wondering how that workshe.  Does the core just match based on the end point location/name?

**Comment by psiinon** [comment]
Yeah, the i18n is just based on the api endpoints. The code generators actually tell you which ones arent present, but you should be able to work them out based on the existing ones.

**Comment by kingthorin** [comment]
Done for allUrls.

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 5]
```diff
@@ -460,6 +468,32 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+			TableHistory tableHistory = extension.getModel().getDb().getTableHistory();
+			List<Integer> ids = Collections.emptyList();
+			try {
+				ids = tableHistory.getHistoryIdsOfHistType(extension.getModel().getSession().getSessionId(),
+						HistoryReference.TYPE_SPIDER, HistoryReference.TYPE_SPIDER_TASK);
+			} catch (DatabaseException e) {
+				throw new ApiException(ApiException.Type.INTERNAL_ERROR, e.getMessage());
+			}
+			for (Integer id : ids) {
+					try {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: The try-catch statement is not correctly indented (it has one more level than it should).

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 6]
```diff
@@ -460,6 +468,32 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+					urlSet.add(
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: This could be tweaked to immediately add the `url` into the `resultUrls` (the `add` method returns  true if it object was added).

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: You mean wrap `resultUrls.addItem` around `urlSet.add`, because it won't do anything if urlSet.add doesn't add it (i.e.: it isn't unique)?

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java (hunk 5 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Thanks, will touch that up.

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java (hunk 6 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: I mean something like:
```Java
if (urlSet.add(url)) {
    resultUrls.addItem(new ApiResponseElement("url", url));
}
```
If it was able to add to the Set it means it's unique and can be added to the results.

File: src/org/zaproxy/zap/extension/spider/SpiderAPI.java [hunk 7]
```diff
@@ -460,6 +468,33 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+				try {
+					url = tableHistory.read(id.intValue()).getHttpMessage().getRequestHeader().getURI().toString();
+					if (urlSet.add(url)) {
+						resultUrls.addItem(new ApiResponseElement("url", url));
+					}
+				} catch (HttpMalformedHeaderException | DatabaseException e) {
+					throw new ApiException(ApiException.Type.INTERNAL_ERROR, e.getMessage());
+				}
+			}
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Since urlSet is only used to enforce uniqueness should I be emptying it here? `urlSet=Collections.emptySet()` or will the normal GC take care of it?

Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: GC will take care of it.

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Great, this one done then? :smiley:

Comment (thc202) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Actually, noticed a potential issue, we should also test that the `RecordHistory` is not `null` before using it, that can happen if one of the entries is deleted while we are iterating the IDs (likely to happen for `TYPE_SPIDER_TASK`, since those are deleted once no longer needed).

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/spider/SpiderAPI.java]: Ok, will address that.

**Comment by kingthorin** [comment]
Done

**Comment by thc202** [comment]
Looks good to me, thank you!

**Comment by psiinon** [comment]
Looks good to me too. Might be nice to have a paging ption as well, as per the history urls, but they can be optional params added later :) Merging now...
